pkg/kubelet/kuberuntime/kuberuntime_container_linux.go

```diff
@@ -89,6 +89,20 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerConfig(container *v1.C
 
 	lc.Resources.HugepageLimits = GetHugepageLimitsFromResources(container.Resources)
 
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwapEnabled) {
+		// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
+		// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
+		switch m.memorySwapBehavior {
+		case "UnlimitedSwap":
+			// -1 = unlimited swap
+			lc.Resources.MemorySwapLimitInBytes = -1
+		default:
+			// memorySwapLimit = total permitted memory+swap; if equal to memory limit, => 0 swap
+			// Note that if memory limit is 0, memory swap limit is ignored.
+			lc.Resources.MemorySwapLimitInBytes = lc.Resources.MemoryLimitInBytes
```

Comment:
So this change is deliberate, to make the CRI behave like dockershim:

https://github.com/kubernetes/kubernetes/blob/24b716673caae31f070b06a337bc12c97ff1d4cb/pkg/kubelet/dockershim/helpers_linux.go#L115-L117

And yeah, the assumption is that there won't be any significant swap usage with these controls set.

[From the KEP:](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#just-set---fail-swap-onfalse)

> Dockershim currently sets swap available for workloads to 0. The CRI does not restrict it at all. This inconsistency makes it difficult or impossible to use swap in production, particularly if a user wants to restrict workloads from using swap when using the CRI rather than dockershim.

I could update the KEP to use a different name if you think "NoSwap" is misleading and we get general agreement. I can also improve the documentation to match.

Reply:
(name change is done)

Reply:
KEP update is https://github.com/kubernetes/enhancements/pull/2794